### DISCUSSION TITLE
Enqueue tracks using harbor.http.register example from page 154 liquidsoap book gets error 5

### DISCUSSION THREAD

**Discussion by jascco** [savonet/liquidsoap#3900] [Q&A] [answered]
Hello,
This is my first posting so hopefully I'm providing pertinent information asking for some help. I just built a new server using Ubuntu 24.04, nginx, mariadb, php8.3, shoutcast and liquidsoap 2.2.4. Everything installed using apt install except shoutcast. My previous server was Ubuntu 20.04 with liquidsoap 2.0.2. On that server I was using server.telnet() to send commands to liquidsoap. On Ubuntu 24.04 the command telnet localhost 2007 fails with connection refused error. I'm trying to learn using server.harbor instead of telnet as that seems like a better approach.

Here is my code based on liquidsoap bokk example on page 154. I don't know how to fix the error. Thank you for any help - it would be most appreciated.

```
server.harbor(port=8007, uri="/telnet")
songRequests = request.queue()
def playRequest(~protocol, ~headers, ~data, uri)
	log.important("Serving uri: #{uri}")
	let (uri, args) = url.split(uri)
	fname = args["file"]
	title = args["title"]
	if file.exists(fname) then
		songRequests.push.uri("annotate:title=#{string.quote(title)}:#{fname}")
		log.important("User request pushed into request queue")
		http.response(data="Request pushed.")
	else
		http.response(code=404, data="Invalid file.")
	end
end
harbor.http.register("/play", playRequest)
```

`liquidsoap --check castle-alpine.liq` produces the following error:
```
At castle-alpine.liq, line 352, char 2-14:
		songRequests.push.uri("annotate:title=#{string.quote(title)}:#{fname}")

Error 5: 
this value has no method `push`
  Its type is source(audio=pcm('B), 'A)
.{
  buffered : () -> [string * float],
  cross_duration : () -> float,
  duration : () -> float,
  elapsed : () -> float,
  fallible : bool,
  id : () -> string,
  is_active : () -> bool,
  is_ready : () -> bool,
  is_up : () -> bool,
  last_metadata : () -> [string * string]?,
  log : 
  {level : (() -> int?).{set : (int) -> unit}
  },
  on_metadata : ((([string * string]) -> unit)) -> unit,
  on_shutdown : ((() -> unit)) -> unit,
  on_track : ((([string * string]) -> unit)) -> unit,
  on_wake_up : ((() -> unit)) -> unit,
  remaining : () -> float,
  seek : (float) -> float,
  self_sync : () -> bool,
  skip : () -> unit,
  time : () -> float
}
where
  'A is a set of tracks to be muxed into a source and a set of internal tracks.
```

**Comment by toots** [comment, marked as answer]
Hi @jascco. I'm not sure why liquidsoap returns this error. Is that the full script? The lines don't seem to match. It's possible that t the variable is redefined, for instance, you are doing:

```liquidsoap
songRequests = request.queue()
songRequests = mksafe(songRequests)
```
In this case, the new definition would hide the one from `request.queue` and would not have its methods.

BTW, the harbor API has been updated in `2.2.x` and is easier to use. You code should now look like this:
```liquidsoap
songRequests = request.queue()

def playRequest(req, res) =
  uri = req.path
  args = req.query

  log.important(
    "Serving uri: #{uri}"
  )

  fname = args["file"]
  title = args["title"]

  if
    file.exists(fname)
  then
    songRequests.push.uri("annotate:title=#{string.quote(title)}:#{fname}")
    log.important(
      "User request pushed into request queue"
    )
    res.data(
      "Request pushed."
    )
  else
    res.data(
      "Invalid file."
    )
    res.status_code(404)
  end
end

harbor.http.register(port=8007, method="GET", "/play", playRequest)
```

**Reply by vitoyucepi** [reply]
Hi @toots,
Is it safe to push the file without filtering user input?

**Reply by jascco** [reply]
Thank you for the prompt response. The entire script is much larger. I didn't have songRequests = mksafe(songRequests) but I did have songRequests = crossfade(id="requestfade",fade_out=3.0, fade_in=3.0, duration=5.0, songRequests)

I didn't realize this would hide the methods from the original songRequests = request.queue()
I've learned something new - thank you.

I used your new harbor code and it works perfectly! Thank you once again.

Related to the question from vitoyurcepi 
In php I always put input from $_GET or $_POST through a function testInput() The code is below. Is there an equivalent way to do this in liquidsoap?

function testInput($data) {
$data = trim($data); // remove extraneous characters from beginning and end of string
$data = stripslashes($data); // remove backslashes escape characters
$data = htmlspecialchars($data); // convert html characters into sequences to prevent browser from rendering them
return $data;
} // end of function testInput($data)
